scripts/support/zephyr_flash_debug.py

```diff
@@ -177,71 +312,77 @@ def create_from_env(debug):
-            extra_init_args = self.extra_init.split()
+        if command in {'flash', 'debug'}:
+            self.flash_debug(command, **kwargs)
+        else:
+            self.debugserver(**kwargs)
 
-        config = path.join(self.zephyr_base, 'boards', self.arch,
-                           self.board_name, 'support', 'openocd.cfg')
+    def flash_debug(self, command, **kwargs):
+        config = kwargs['openocd-cfg']
 
-        helper_cmd = ([self.openocd] +
-                      search_args +
+        server_cmd = (self.openocd_cmd +
                       ['-f', config] +
-                      extra_init_args +
+                      self.extra_init +
                       ['-c', 'tcl_port {}'.format(self.tcl_port),
                        '-c', 'telnet_port {}'.format(self.telnet_port),
                        '-c', 'gdb_port {}'.format(self.gdb_port),
                        '-c', 'init',
                        '-c', 'targets',
                        '-c', 'halt'])
-        helper = subprocess.Popen([self.python, 'arc-helper.py'] +
-                                  helper_cmd)
 
         tui_arg = []
         if self.tui is not None:
             tui_arg = [self.tui]
 
+        continue_arg = []
+        if command == 'flash':
+            continue_arg = ['-ex', 'c']
+
         gdb_cmd = ([self.gdb] +
                    tui_arg +
                    ['-ex', 'target remote :{}'.format(self.gdb_port),
-                    '-ex', 'load',
-                    '-ex', 'c',
-                   self.elf])
-        previous = signal.signal(signal.SIGINT, signal.SIG_IGN)
-        try:
-            check_call(gdb_cmd, self.debug)
-        finally:
-            signal.signal(signal.SIGINT, previous)
-            helper.terminate()
-            helper.wait()
+                    '-ex', 'load'] +
+                   continue_arg +
+                   [self.elf])
+
+        self.run_server_and_client(server_cmd, gdb_cmd)
+
+    def debugserver(self, **kwargs):
+        config = kwargs['openocd-cfg']
+        cmd = (self.openocd_cmd +
+               ['-f', config,
+                '-c', 'init',
+                '-c', 'targets',
+                '-c', 'reset halt'])
+        check_call(cmd, self.debug)
 
 
-class BossacBinaryFlasher(ZephyrBinaryFlasher):
-    '''Flasher front-end for bossac.'''
+class BossacBinaryRunner(ZephyrBinaryRunner):
+    '''Runner front-end for bossac.'''
 
     def __init__(self, bin_name, bossac='bossac', debug=False):
-        super(BossacBinaryFlasher, self).__init__(debug=debug)
+        super(BossacBinaryRunner, self).__init__(debug=debug)
         self.bin_name = bin_name
         self.bossac = bossac
 
-    def replaces_shell_script(shell_script):
-        return shell_script == 'bossa-flash.sh'
+    def replaces_shell_script(shell_script, command):
+        return command == 'flash' and shell_script == 'bossa-flash.sh'
 
-    def create_from_env(debug):
+    def create_from_env(command, debug):
```

Comment:
Since you are cleaning all things up, can you look at adding in when I did with PR #4502

Reply:
Sounds good! That class was a pretty primitive one.